editor/src/components/canvas/ui-jsx-canvas.tsx

```diff
@@ -267,10 +283,68 @@ export const UiJsxCanvas = betterReactMemo(
       clearErrors()
     }
 
+    // TODO after merge requireFn can never be null
     if (requireFn != null) {
       const customRequire = React.useCallback(
-        (importOrigin: string, toImport: string) => requireFn(importOrigin, toImport, false),
-        [requireFn],
+        (importOrigin: string, toImport: string) => {
+          const filePathResolveResult = resolve(importOrigin, toImport)
+          const resolvedParseSuccess: Either<string, MapLike<any>> = flatMapEither(
+            (resolvedFilePath) => {
+              const projectFile = getContentsTreeFileFromString(projectContents, resolvedFilePath)
+              if (isTextFile(projectFile) && isParseSuccess(projectFile.fileContents.parsed)) {
+                const { scope } = createExecutionScope(
+                  resolvedFilePath,
+                  customRequire,
+                  mutableContextRef,
+                  topLevelComponentRendererComponents,
+                  projectContents,
+                  uiFilePath,
+                  transientFileState,
+                  base64FileBlobs,
+                  hiddenInstances,
+                  metadataContext,
+                  shouldIncludeCanvasRootInTheSpy,
+                )
+                const exportsDetail = projectFile.fileContents.parsed.exportsDetail
+                let filteredScope: MapLike<any> = {}
+                for (const s of Object.keys(scope)) {
+                  if (s in exportsDetail.namedExports) {
+                    filteredScope[s] = scope[s]
+                  } else if (s === exportsDetail.defaultExport?.name) {
+                    filteredScope[s] = scope[s]
+                  }
+                }
+                return right(filteredScope)
+              } else {
+                return left(`File ${resolvedFilePath} is not a ParseSuccess`)
+              }
+            },
+            filePathResolveResult,
+          )
+          return foldEither(
+            () => {
+              // We did not find a ParseSuccess, fallback to standard require Fn
+              return requireFn(importOrigin, toImport, false)
+            },
+            (scope) => {
+              // Return an artificial exports object that contains our ComponentRendererComponents
+              return scope
+            },
+            resolvedParseSuccess,
+          )
+        },
+        // TODO I don't like projectContents and transientFileState here because that means dragging smth on the Canvas would recreate the customRequire fn
```

Comment:
Danger Will Robinson, performance?

Reply:
I'm assuming this is part of the next load of work?

Reply:
well this is a placeholder for discussion – right now we do not have a better idea than this, and we are actually re-creating the global requireFn for any file change as well, so I think this is definitely not a performance _regression_ just an area to improve